pkg/controller/snapshotter/snapshotter.go

```diff
@@ -347,204 +345,310 @@ func (vs *volumeSnapshotter) deleteSnapshot(spec *v1.PersistentVolumeSpec, sourc
-					err = vs.restClient.Post().
-						Resource(crdv1.VolumeSnapshotDataResourcePlural).
-						Namespace(v1.NamespaceDefault).
-						Body(snapshotData).
-						Do().Into(&result)
-					if err != nil {
-						// Re-Try it as errors writing to the API server are common
-						time.Sleep(createVolumeSnapshotDataInterval)
-					} else {
-						break
-					}
-				}
+	var tags *map[string]string = nil
+	if !bFoundMeta {
+		glog.Infof("createSnapshot: Metadata not found. Create metadata for snapshot %s.", snapshotName)
+		tags, err = vs.updateVolumeSnapshotMetadata(snapshot)
+		if err != nil {
+			return fmt.Errorf("Failed to update metadata for volume snapshot %s: %q", snapshotName, err)
+		}
+	}
 
-				if err != nil {
-					glog.Errorf("Error creating the VolumeSnapshotData %s: %v", snapshotName, err)
-					// Don't proceed to create snapshot using the plugin due to error creating
-					// VolumeSnapshotData
-					return fmt.Errorf("Failed to create the VolumeSnapshotData %s for snapshot %s", snapName, snapshotName)
-				}
-			}
+	if snapshotDataSource == nil {
+		glog.Infof("createSnapshot: Snapshot not created yet. Creating snapshot %s through the plugin ...", snapshotName)
+		pv, err := vs.getPVFromVolumeSnapshot(snapshotName, snapshot)
+		if err != nil {
+			return err
+		}
+		snapshotDataSource, snapStatus, err = vs.takeSnapshot(pv, tags)
+		if err != nil || snapshotDataSource == nil {
+			return fmt.Errorf("Failed to take snapshot of the volume %s: %q", pv.Name, err)
+		}
+	}
 
-			// Update the VolumeSnapshot object with status. This should bind VolumeSnapshotData to VolumeSnapshot object
-			snapConditions := []crdv1.VolumeSnapshotCondition{
-				{
-					Type:               crdv1.VolumeSnapshotConditionPending,
-					Status:             v1.ConditionTrue,
-					Message:            "Snapshot created succsessfully",
-					LastTransitionTime: metav1.Now(),
-				},
-			}
-			_, err = vs.UpdateVolumeSnapshot(snapshotName, &snapConditions, true)
-			if err != nil {
-				glog.Errorf("Error updating volume snapshot %s: %v", snapshotName, err)
-				// NOTE(xyang): Return error if failed to update VolumeSnapshot after
-				// create snapshot request is sent to the plugin and VolumeSnapshotData is updated
-				return fmt.Errorf("Failed to update VolumeSnapshot for snapshot %s", snapshotName)
-			}
+	glog.Infof("createSnapshot: Update status for VolumeSnapshot object %s.", snapshotName)
+        _, err = vs.UpdateVolumeSnapshot(snapshotName, snapStatus)
+        if err != nil {
+                glog.Errorf("createSnapshot: Error updating volume snapshot %s: %v", snapshotName, err)
+                return fmt.Errorf("Failed to update VolumeSnapshot for snapshot %s", snapshotName)
+        }
 
-			return nil
+	glog.Infof("createSnapshot: Check if VolumeSnapshotData object already created for VolumeSnapshot %s.", snapshotName)
+	snapshotDataObj := vs.getSnapshotDataFromSnapshotName(snapshotName)
```

Comment:
I think snapshotData object must be nil at here, don't need to check

Reply:
Ok